middleman_protocol/tests/test_message.py

```diff
@@ -0,0 +1,207 @@
+        (AuthenticationChallengeFrame, b'random_bytes'),
+        (AuthenticationResponseFrame,  b'TODO'),
+    ])
+    def test_that_sending_message_over_tcp_socket_should_preserve_original_data(
+        self,
+        middleman_message_type,
+        payload,
+    ):
+        middleman_message = middleman_message_type(payload, self.request_id)
+
+        with closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as server_socket:
+            server_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, True)
+            with closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as client_socket:
+                client_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, True)
+                server_socket.bind(('127.0.0.1', 8001))
+                server_socket.listen(1)
+
+                client_socket.connect(('127.0.0.1', 8001))
+
+                (connection, _address) = server_socket.accept()
+
+                send_over_stream(
+                    connection=client_socket,
+                    raw_message=middleman_message,
+                    private_key=CONCENT_PRIVATE_KEY
+                )
+                deserialized_message = next(receive_frame(connection=connection, public_key=CONCENT_PUBLIC_KEY))
+
+        assertpy.assert_that(deserialized_message).is_instance_of(middleman_message_type)
+        assertpy.assert_that(deserialized_message.payload).is_instance_of(type(payload))
+        assertpy.assert_that(deserialized_message.payload).is_equal_to(payload)
+
+    def test_that_serializing_different_golem_message_middleman_messages_should_keep_part_of_header_the_same(self):
+        message_1 = GolemMessageFrame(
+            ServiceRefused(reason=ServiceRefused.REASON.InvalidRequest),
+            self.request_id
+        ).serialize(
+            private_key=CONCENT_PRIVATE_KEY
+        )
+        message_2 = GolemMessageFrame(Ping(), self.request_id).serialize(private_key=CONCENT_PRIVATE_KEY)
+
+        assertpy.assert_that(
+            message_1[FRAME_SIGNATURE_BYTES_LENGTH:FRAME_PAYLOAD_STARTING_BYTE]
+        ).is_equal_to(
+            message_2[FRAME_SIGNATURE_BYTES_LENGTH:FRAME_PAYLOAD_STARTING_BYTE]
+        )
+
+    @pytest.mark.parametrize(('middleman_message_type', 'payload'), [
+        (GolemMessageFrame,            Ping()),
+        (ErrorFrame,                   (111, 'error_message')),
+        (AuthenticationChallengeFrame, b'random_bytes'),
+        (AuthenticationResponseFrame,  b'TODO'),
+    ])
+    def test_that_serializing_and_deserializing_message_with_wrong_signature_should_raise_exception(
+        self,
+        middleman_message_type,
+        payload,
+    ):
+        middleman_message = middleman_message_type(payload, self.request_id)
+        raw_message = middleman_message.serialize(private_key=CONCENT_PRIVATE_KEY)
+
+        first_byte = raw_message[1] if raw_message[1] != 0 else 1
+        malformed_raw_message = bytes(bytearray([first_byte - 1])) + raw_message[1:]
```

Comment:
hm... it seems that if `raw_message[1] == 0 `, then `first_byte = 1` and `bytes(bytearray([first_byte - 1]))` will produce `0` again, meaning that `raw_message == malformed_raw_message` (so the message **is not malformed**)